### PR TITLE
cluster_config: init cluster variables on startup

### PR DESCRIPTION
**Pull request by ykaliuta**
Cluster configuration is supposed to be the same during operator pod
lifetime. There is no point to detect it on every reconcile cycle
keeping in mind that it can causes many api requests.

Do the initialization on startup. Keep GetOperatorNamespace()
returning error since it defines some logic in the DSCInitialization
reconciler.

Automatically called init() won't work here due to need to check
error of the initialization.

Leave GetDomain() as is since it uses OpenshiftIngress configuration
which is created when DSCInitialization instantiated.

Log cluster configuration on success.

Many thanks for submitting your Pull Request ❤️!

Please make sure that your PR meets the following requirements:

- [ ] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md)
- [ ] Pull Request contains description of the issue
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request

## Merge criteria:

- [ ] Have a meaningful commit messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: pkg/cluster/cluster_config.go [hunk 1]
```diff
@@ -22,13 +23,68 @@ import (
 // +kubebuilder:rbac:groups="config.openshift.io",resources=ingresses,verbs=get
 
 type Platform string
+
```
Comment (ykaliuta) [inline pkg/cluster/cluster_config.go:26]: Unrelated change :( But will not update until other changes needed due to the test instability

File: pkg/cluster/cluster_config.go [hunk 2]
```diff
@@ -21,6 +22,69 @@ import (
...
+	if err != nil {
+		return err
+	}
+
+	clusterConfig.Release, err = getRelease(ctx, cli)
+	if err != nil {
+		return err
+	}
+
+	printClusterConfig(log)
```
[the remark this answers is not in the document]
Comment (ykaliuta) [reply pkg/cluster/cluster_config.go:61]: I was changing format in the meantime and did not want to pollute the main function with the details. But if you like I can inline it.

**Comment by lburgazzoli** [comment]
LGTM

File: pkg/cluster/cluster_config.go [hunk 3]
```diff
@@ -21,6 +22,69 @@ import (
...
+}
+
+func printClusterConfig(log logr.Logger) {
+	log.Info("Cluster config",
+		"Namespace", clusterConfig.Namespace,
+		"Platform", clusterConfig.Platform,
+		"Release", clusterConfig.Release)
+}
+
+func GetOperatorNamespace() (string, error) {
```
Comment (zdtsw) [inline pkg/cluster/cluster_config.go:73]: so 
`getOperatorNamespace()` call with real logic and return retrieve value
`GetOperatorNamespace(...) `return clusterConfig.Namespace


similar to 
`getPlatform() ` and `getRelease()` call logic return value
`GetPlatform(...)` and `GetRelease(...)` return filed's value

GetRelease(...) indirectly rely on `getPlatform() `

File: pkg/cluster/cluster_config.go [hunk 4]
```diff
@@ -21,6 +22,69 @@ import (
...
+	Platform  Platform
+	Release   Release
+}
+
+// Init initializes cluster configuration variables on startup
+// init() won't work since it is needed to check the error.
+func Init(ctx context.Context, cli client.Client, log logr.Logger) error {
+	var err error
+
+	clusterConfig.Namespace, err = getOperatorNamespace()
```
[the remark this answers is not in the document]
Comment (ykaliuta) [reply pkg/cluster/cluster_config.go:45]: in configureAlertManager and in watchMonitoringSecretResource the error causes return. In other places where the value compared "" does not change the flow. I did not revisit callees' logic. It may make sense.

> should this not be fatal => os.Exit(1) ?

it's not fatal. Do you mean it should be?

[the remark this answers is not in the document]
Comment (zdtsw) [reply pkg/cluster/cluster_config.go:45]: i am thinking it should be fatal.
for this comments `// not fatal, fallback to ""`

[the remark this answers is not in the document]
Comment (ykaliuta) [reply pkg/cluster/cluster_config.go:45]: > i am thinking it should be fatal. for this comments `// not fatal, fallback to ""`

It changes behaviour. Mean, we continued to work if no environment variable or cannot read namespace from the filesystem, but if it's fatal we don't.

On the other hand with the environment variable corner cases (like tests) are covered.

[the remark this answers is not in the document]
Comment (zdtsw) [reply pkg/cluster/cluster_config.go:45]: i understand.
1. env
2. check /var/...
3. log as error but continue
but the consequence on (3) will leave to later problem, so if it should os.Exit(1) directly in main process or not.
e.g i can understand someone build the whole product without having (1) set, but the reason can skip (2) is something really wrong with the pod. 

but this is a question out of the scope of your PR.
the "" has been there, we do not need to change anything here

**Review by zdtsw — COMMENTED** [review]
LGTM

**Comment by ykaliuta** [comment]
Can I get review before it needs rebase again? :) Added change with preserving rootCtx till mgr.Start.

File: main.go [hunk 5]
```diff
@@ -127,7 +128,8 @@ func main() { //nolint:funlen,maintidx
 	ctrl.SetLogger(logger.ConfigLoggers(logmode))
 
 	// root context
-	ctx := ctrl.SetupSignalHandler()
+	rootCtx := ctrl.SetupSignalHandler()
+	ctx := logf.IntoContext(rootCtx, setupLog)
```
Comment (zdtsw) [inline main.go:132]: not sure i understand why for this change: 
so the old code is, using rootCtx everywhere
the new code is, pass the new ctx which is `context.WithValue(rootCtx, contextKey{}, setupLog) ` instead except to start manager.

Comment (ykaliuta) [reply main.go:132]: the new  code should pass the original context without setuplog since I found it a bit incorrect to pass log.WithName("setup") further. Did I do a mistake?

Comment (zdtsw) [reply main.go:132]: so at the beginning, we do `setupLog = ctrl.Log.WithName("setup")`
then later e.g 
```
if err = (&dscictrl.DSCInitializationReconciler{
		Client:                mgr.GetClient(),
		Scheme:                mgr.GetScheme(),
		Log:                   logger.LogWithLevel(ctrl.Log.WithName(operatorName).WithName("controllers").WithName("DSCInitialization"), logmode),
		Recorder:              mgr.GetEventRecorderFor("dscinitialization-controller"),
		ApplicationsNamespace: dscApplicationsNamespace,
	}).SetupWithManager(ctx, mgr); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "DSCInitiatlization")
		os.Exit(1)
	}
```
call with setupLog when problem to start contoller.
context is not used for for logging `"unable to create controller", "controller", "DSCInitiatlization".`
or?

Comment (ykaliuta) [reply main.go:132]: you mean ctx in .SetupWithManager(ctx, mgr)? It is not used at all IIUC, I planned patch to remove it.

Comment (zdtsw) [reply main.go:132]: yeah, but what is "the new code should pass the original context without setuplog"? 

because from this change, except the `mgr.Start(rootCtx`) is kept to use the root context (regardless what the name of variable is now), everywhere else is using the new one (logf.IntoContext(rootCtx, setupLog)). so i think the way e.g  `upgrade.CreateDefaultDSCI(ctx,...) ` is starting to have "setup" with this PR vs it did not before the PR? which is  "the new code should pass the original context with setuplog" :thinking:

Comment (ykaliuta) [reply main.go:132]: Yes, that's the point. When (if) we switch to logger in context, everything before mgr.Start() is setup and will have "setup" logger, `upgrade.CreateDefaultDSCI()` as well. But `Reconcile` is not setup anymore. Of course, it can just ignore the logger from ctx, but I wanted to make it explicit.

Comment (ykaliuta) [reply main.go:132]: Ignore my last comment :) Let me think

Comment (ykaliuta) [reply main.go:132]: So `upgrade.CreateDefaultDSCI(ctx,...)` as executed by manager after `Start()` will have `ctx` without "setup" logger. Basically, what it uses now getting logger directly from the controller.

Comment (ykaliuta) [reply main.go:132]: Actually, controller framework passes its own logger with the context, so it's not needed to save rootCtx.

Comment (ykaliuta) [reply main.go]: @zdtsw I reverted that change. Any more concerns?

**Review by zdtsw — APPROVED** [review]
